### ISSUE TITLE
quoted-strings false positive on strings containing brackets

### ISSUE THREAD

**Issue by sxlijin** [adrienverge/yamllint#516]
When using brackets to define lists, string literals in that list can't contain brackets; 

```
$ cat foo.yaml
++ 2022-11-04T11:24:52-0700 ++ cat foo.yaml
# invalid
field: [string[bracket]]
```

but when you include brackets:

```
$ cat foo.yaml
++ 2022-11-04T11:24:52-0700 ++ cat foo.yaml
# valid
field: ["string[bracket]"]
```

yamllint complains about the string being quoted:

```
$ trunk check --filter=yamllint foo.yaml
++ 2022-11-04T11:24:54-0700 ++ trunk check --filter=yamllint foo.yaml
Checking 100% [=======================================================================================================================================================>]  5/5  0.3s

  ISSUES

foo.yaml:2:9
 2:9  high  string value is redundantly quoted with any quotes  yamllint/quoted-strings

✖ 1 blocking issue
```

**Comment by adrienverge** [comment]
Hello @sxlijin, thanks for reporting this.

A simpler reproduction is:
```shell
$ yamllint -d 'rules: {quoted-strings: {required: only-when-needed}}' - <<<'field: [string[bracket]]'
  1:15      error    syntax error: expected ',' or ']', but got '[' (syntax)

$ yamllint -d 'rules: {quoted-strings: {required: only-when-needed}}' - <<<'field: ["string[bracket]"]'
  1:9       error    string value is redundantly quoted with any quotes  (quoted-strings)
```

@ruipinge you contributed `required: only-when-needed` in #235, would you like to have a look into this?

In the meantine, a YAML block sequence can be used to workaround the problem:
```yaml
field: 
- string[bracket]
```

**Comment by robherring** [comment]
There's also a related case with commas in bracketed sequences:

`
$ yamllint -d 'rules: {quoted-strings: {required: only-when-needed}}' - <<<'field: [ "foo,bar" ]'
stdin
  1:10      error    string value is redundantly quoted with any quotes  (quoted-strings)
`
